Add NCCL_ASYNC_ERROR_HANDLING to docs

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46856 Add NCCL_ASYNC_ERROR_HANDLING to docs**

Add reference to NCCL_ASYNC_ERROR_HANDLING in the pytorch docs,
similar to how NCCL_BLOCKING_WAIT is curently described.

Differential Revision: [D24541822](https://our.internmc.facebook.com/intern/diff/D24541822/)

osalpekar committed:
Add NCCL_ASYNC_ERROR_HANDLING to docs

Add reference to NCCL_ASYNC_ERROR_HANDLING in the pytorch docs,
similar to how NCCL_BLOCKING_WAIT is curently described.

Differential Revision: [D24541822](https://our.internmc.facebook.com/intern/diff/D24541822/)

[ghstack-poisoned]

osalpekar committed:
Update on "Add NCCL_ASYNC_ERROR_HANDLING to docs"


Add reference to NCCL_ASYNC_ERROR_HANDLING in the pytorch docs,
similar to how NCCL_BLOCKING_WAIT is curently described.

Differential Revision: [D24541822](https://our.internmc.facebook.com/intern/diff/D24541822/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -393,7 +393,12 @@ def init_process_group(backend,
+            or ``NCCL_ASYNC_ERROR_HANDLING`` is set to 1. When
+            ``NCCL_BLOCKING_WAIT`` is set, this is the duration for which the
+            process will block and wait for collectives to complete before
+            throwing an exception. When ``NCCL_ASYNC_ERROR_HANDLING`` is set,
```
What if both `NCCL_ASYNC_ERROR_HANDLING` and `NCCL_ASYNC_ERROR_HANDLING` are set to 1? Or is this case not allowed?

osalpekar replied:
This should be avoided, but NCCL_BLOCKING_WAIT will catch the error first and throw an exception.

jiayisuse replied:
* Maybe we can add a comment ``NCCL_BLOCKING_WAIT`` is set to 1. or ``NCCL_ASYNC_ERROR_HANDLING`` is set to 1 **(these two configurations are exclusive)**

* which the process **(host thread)** will block and

jiayisuse commented on torch/distributed/distributed_c10d.py:
```diff
@@ -393,7 +393,12 @@ def init_process_group(backend,
+            process will block and wait for collectives to complete before
+            throwing an exception. When ``NCCL_ASYNC_ERROR_HANDLING`` is set,
+            this is the duration after which collectives will be aborted
+            asynchronously and the process will crash.
```
will crash **with timeout message telling the timeout collective call, process ID, thread ID and etc.**

osalpekar replied:
The detailed logs are not cherry-picked to 1.7 so we will throw an `std::runtime_error("NCCL Operation Timed Out")` and then crash (user can get pid and SIGABRT signal from the stderr).

osalpekar committed:
Update on "Add NCCL_ASYNC_ERROR_HANDLING to docs"


Add reference to NCCL_ASYNC_ERROR_HANDLING in the pytorch docs,
similar to how NCCL_BLOCKING_WAIT is curently described.

Differential Revision: [D24541822](https://our.internmc.facebook.com/intern/diff/D24541822/)

[ghstack-poisoned]

osalpekar committed:
Update on "Add NCCL_ASYNC_ERROR_HANDLING to docs"


Add reference to NCCL_ASYNC_ERROR_HANDLING in the pytorch docs,
similar to how NCCL_BLOCKING_WAIT is curently described.

Differential Revision: [D24541822](https://our.internmc.facebook.com/intern/diff/D24541822/)

[ghstack-poisoned]

osalpekar committed:
Update on "Add NCCL_ASYNC_ERROR_HANDLING to docs"


Add reference to NCCL_ASYNC_ERROR_HANDLING in the pytorch docs,
similar to how NCCL_BLOCKING_WAIT is curently described.

Differential Revision: [D24541822](https://our.internmc.facebook.com/intern/diff/D24541822/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -393,7 +393,13 @@ def init_process_group(backend,
+            process will block and wait for collectives to complete before
+            throwing an exception. When ``NCCL_ASYNC_ERROR_HANDLING`` is set,
+            this is the duration after which collectives will be aborted
+            asynchronously and the process will crash. Only one of these two
```
Enhance this comment to mention when to use which argument. Ex: `NCCL_BLOCKING_WAIT` will provide errors to the user which can be caught and handled but due to the blocking nature it has a perf overhead. `NCCL_ASYNC_ERROR_HANDLING` doesn't have any perf overhead, but crashes the process on errors without giving the user any chance to handle errors. This is done since CUDA execution is async and it is no longer safe to continue executing user code since failed async NCCL operations might result in subsequent CUDA operations to run on corrupted data.

osalpekar committed:
Update on "Add NCCL_ASYNC_ERROR_HANDLING to docs"


Add reference to NCCL_ASYNC_ERROR_HANDLING in the pytorch docs,
similar to how NCCL_BLOCKING_WAIT is curently described.

Differential Revision: [D24541822](https://our.internmc.facebook.com/intern/diff/D24541822/)

[ghstack-poisoned]
